### PR TITLE
[NVPTX] Add `-march=general` option to mirror default configuration

### PR DESCRIPTION
**Pull request by oraluben**
This PR adds `-march=generic` support for the NVPTX backend. This fulfills a TODO introduced in #79873.

With this PR, users can explicitly request the "default" CUDA architecture, which makes sure that no specific architecture is specified.

This PR does not address any compatibility issues between different CUDA versions.

### CODE REVIEW

**Comment by oraluben** [comment]
@jhuber6 @Artem-B You might want to check if this LGTY :)

**Review by jhuber6 — CHANGES_REQUESTED** [review]
Thanks for looking at this. When the user compiles with `-march=xyz` it introduces a lot of subtarget specific metadata intro the output IR. The purpose of the original patch was to keep `-target-cpu` unset in cases where `-march=xyz` was not passed in. The expected semantics here is that `-march=sm_52 -march=generic` will override `-march=sm_52` and result in no `-target-cpu` being set just like if you didn't pass `-march` at all.

File: clang/test/Driver/cuda-cross-compiling.c [hunk 1]
```diff
@@ -85,6 +90,6 @@
 // MISSING: error: Must pass in an explicit nvptx64 gpu architecture to 'nvlink'
 
 // RUN: %clang -target nvptx64-nvidia-cuda -flto -c %s -### 2>&1 \
-// RUN:   | FileCheck -check-prefix=GENERIC %s
+// RUN:   | FileCheck -check-prefix=COMPILE %s
```
[the remark this answers is not in the document]
Comment (oraluben) [reply clang/test/Driver/cuda-cross-compiling.c:94]: I see, thanks for clarifying!

**Comment by jhuber6** [comment]
FWIW I think you can kind of do this with `-march=sm_52 -march=` to just set it to empty.

**Review by jhuber6 — APPROVED** [review]
LG, thanks for the patch.

File: clang/lib/Driver/ToolChains/Cuda.cpp [hunk 2]
```diff
@@ -750,10 +750,11 @@ NVPTXToolChain::TranslateArgs(const llvm::opt::DerivedArgList &Args,
     if (!llvm::is_contained(*DAL, A))
       DAL->append(A);
 
-  // TODO: We should accept 'generic' as a valid architecture.
   if (!DAL->hasArg(options::OPT_march_EQ) && OffloadKind != Action::OFK_None) {
     DAL->AddJoinedArg(nullptr, Opts.getOption(options::OPT_march_EQ),
                       CudaArchToString(CudaArch::CudaDefault));
+  } else if (DAL->getLastArgValue(options::OPT_march_EQ) == "generic") {
```
Comment (jhuber6) [inline clang/lib/Driver/ToolChains/Cuda.cpp:756]:
```suggestion
  } else if (DAL->getLastArgValue(options::OPT_march_EQ) == "generic"
                 && OffloadKind == Action::OFK_None) {
```
Ah, forgot, we probably don't want to expose this to CUDA just yet.

**Pull request merged**